File: frontend/lib/api/language.ts
import { z } from "zod";

import { API_URL, ApiError, apiFetch } from "@/lib/api/client";

export const languageCodeSchema = z.enum(["es", "fr"]);
export type LanguageCode = z.infer<typeof languageCodeSchema>;

export const cardTypeSchema = z.enum(["basic", "cloze", "audio"]);
export type CardType = z.infer<typeof cardTypeSchema>;

export const cardDirectionSchema = z.enum(["recognition", "production"]);
export type CardDirection = z.infer<typeof cardDirectionSchema>;

export const reviewStateSchema = z.enum(["new", "learning", "review", "relearning"]);
export type ReviewState = z.infer<typeof reviewStateSchema>;

// ---------------------------------------------------------------------------
// decks
// ---------------------------------------------------------------------------

export const deckSchema = z.object({
  id: z.number(),
  name: z.string(),
  language: languageCodeSchema,
  description: z.string(),
  tags: z.array(z.string()),
  is_system: z.boolean(),
  created_at: z.string(),
  card_count: z.number(),
  due_count: z.number(),
  new_count: z.number(),
});
export type Deck = z.infer<typeof deckSchema>;

export type DeckIn = {
  name: string;
  language: LanguageCode;
  description?: string;
  tags?: string[];
};

export function listDecks(): Promise<Deck[]> {
  return apiFetch("/language/decks", z.array(deckSchema));
}

export function createDeck(body: DeckIn): Promise<Deck> {
  return apiFetch("/language/decks", deckSchema, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updateDeck(
  id: number,
  body: Partial<Omit<DeckIn, "language">>,
): Promise<Deck> {
  return apiFetch(`/language/decks/${id}`, deckSchema, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function deleteDeck(id: number): Promise<void> {
  const res = await fetch(`${API_URL}/language/decks/${id}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new ApiError(res.status, body || res.statusText);
  }
}

// ---------------------------------------------------------------------------
// cards
// ---------------------------------------------------------------------------

export const cardSchema = z.object({
  id: z.number(),
  deck_id: z.number(),
  card_type: cardTypeSchema,
  direction: cardDirectionSchema,
  front: z.string(),
  back: z.string(),
  ipa: z.string(),
  gender: z.string(),
  part_of_speech: z.string(),
  audio_url: z.string(),
  example: z.string(),
  example_translation: z.string(),
  cognate_note: z.string(),
  is_false_friend: z.boolean(),
  source: z.string(),
  source_ref: z.string(),
  lexeme_id: z.number().nullable(),
  conjugation_id: z.number().nullable(),
  tags: z.array(z.string()),
  created_at: z.string(),
  state: reviewStateSchema,
  due: z.string(),
  reps: z.number(),
  lapses: z.number(),
  stability: z.number().nullable(),
});
export type Card = z.infer<typeof cardSchema>;

export type CardIn = {
  deck_id: number;
  front: string;
  back?: string;
  card_type?: CardType;
  direction?: CardDirection;
  ipa?: string;
  gender?: string;
  part_of_speech?: string;
  example?: string;
  example_translation?: string;
  cognate_note?: string;
  tags?: string[];
  enrich?: boolean;
};

export function listCards(deckId: number): Promise<Card[]> {
  return apiFetch(`/language/decks/${deckId}/cards`, z.array(cardSchema));
}

export function createCard(body: CardIn): Promise<Card> {
  return apiFetch("/language/cards", cardSchema, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updateCard(
  id: number,
  body: Partial<Omit<CardIn, "enrich">>,
): Promise<Card> {
  return apiFetch(`/language/cards/${id}`, cardSchema, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export function enrichCard(id: number): Promise<Card> {
  return apiFetch(`/language/cards/${id}/enrich`, cardSchema, { method: "POST" });
}

export async function deleteCard(id: number): Promise<void> {
  const res = await fetch(`${API_URL}/language/cards/${id}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new ApiError(res.status, body || res.statusText);
  }
}

// ---------------------------------------------------------------------------
// study
// ---------------------------------------------------------------------------

export const studyQueueSchema = z.object({
  cards: z.array(cardSchema),
  due_count: z.number(),
  new_count: z.number(),
});
export type StudyQueue = z.infer<typeof studyQueueSchema>;

export const reviewOutSchema = z.object({
  card_id: z.number(),
  state: reviewStateSchema,
  due: z.string(),
  again_soon: z.boolean(),
});
export type ReviewOut = z.infer<typeof reviewOutSchema>;

export function getStudyQueue(params: {
  deckId?: number;
  verbSetId?: number;
  language?: LanguageCode;
  newLimit?: number;
}): Promise<StudyQueue> {
  const search = new URLSearchParams();
  if (params.deckId != null) search.set("deck_id", String(params.deckId));
  if (params.verbSetId != null)
    search.set("verb_set_id", String(params.verbSetId));
  if (params.language) search.set("language", params.language);
  if (params.newLimit != null) search.set("new_limit", String(params.newLimit));
  return apiFetch(`/language/study/queue?${search}`, studyQueueSchema);
}

export function reviewCard(cardId: number, rating: number): Promise<ReviewOut> {
  return apiFetch("/language/study/review", reviewOutSchema, {
    method: "POST",
    body: JSON.stringify({ card_id: cardId, rating }),
  });
}

// ---------------------------------------------------------------------------
// text library + annotations
// ---------------------------------------------------------------------------

export const textAnnotationSchema = z.object({
  id: z.number(),
  text_id: z.number(),
  start_offset: z.number(),
  end_offset: z.number(),
  selected_text: z.string(),
  kind: z.string(),
  color: z.string(),
  note: z.string(),
  translation: z.string(),
  ipa: z.string(),
  gender: z.string(),
  part_of_speech: z.string(),
  cognate_note: z.string(),
  is_false_friend: z.boolean(),
  lexeme_id: z.number().nullable(),
  headword: z.string(),
  form_note: z.string(),
  deck_id: z.number().nullable(),
  flashcard_id: z.number().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type TextAnnotation = z.infer<typeof textAnnotationSchema>;

export const languageTextSummarySchema = z.object({
  id: z.number(),
  title: z.string(),
  language: languageCodeSchema,
  source_type: z.string(),
  source_ref: z.string(),
  tags: z.array(z.string()),
  created_at: z.string(),
  updated_at: z.string(),
  annotation_count: z.number(),
});
export type LanguageTextSummary = z.infer<typeof languageTextSummarySchema>;

export const languageTextDetailSchema = languageTextSummarySchema.extend({
  content: z.string(),
  annotations: z.array(textAnnotationSchema),
});
export type LanguageTextDetail = z.infer<typeof languageTextDetailSchema>;

export type LanguageTextIn = {
  title: string;
  language: LanguageCode;
  content: string;
  source_type?: string;
  source_ref?: string;
  tags?: string[];
};

export type TextAnnotationIn = {
  start_offset: number;
  end_offset: number;
  selected_text?: string;
  kind?: string;
  color?: string;
  note?: string;
  translation?: string;
  ipa?: string;
  gender?: string;
  part_of_speech?: string;
  cognate_note?: string;
  is_false_friend?: boolean;
  headword?: string;
  form_note?: string;
};

export const cardLocationSchema = z.object({
  card_id: z.number(),
  deck_id: z.number(),
  deck_name: z.string(),
});

export const textLookupSchema = z.object({
  selected_text: z.string(),
  translation: z.string(),
  ipa: z.string(),
  gender: z.string(),
  part_of_speech: z.string(),
  cognate_note: z.string(),
  is_false_friend: z.boolean(),
  provider: z.string(),
  headword: z.string(),
  is_inflected: z.boolean(),
  form_note: z.string(),
  lexeme_id: z.number().nullable(),
  existing_cards: z.array(cardLocationSchema),
});
export type TextLookup = z.infer<typeof textLookupSchema>;

export function listLanguageTexts(language?: LanguageCode): Promise<LanguageTextSummary[]> {
  const search = new URLSearchParams();
  if (language) search.set("language", language);
  const suffix = search.toString() ? `?${search}` : "";
  return apiFetch(`/language/texts${suffix}`, z.array(languageTextSummarySchema));
}

export function createLanguageText(body: LanguageTextIn): Promise<LanguageTextDetail> {
  return apiFetch("/language/texts", languageTextDetailSchema, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function getLanguageText(id: number): Promise<LanguageTextDetail> {
  return apiFetch(`/language/texts/${id}`, languageTextDetailSchema);
}

export function updateLanguageText(
  id: number,
  body: Partial<LanguageTextIn>,
): Promise<LanguageTextDetail> {
  return apiFetch(`/language/texts/${id}`, languageTextDetailSchema, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function deleteLanguageText(id: number): Promise<void> {
  const res = await fetch(`${API_URL}/language/texts/${id}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new ApiError(res.status, body || res.statusText);
  }
}

export function createTextAnnotation(
  textId: number,
  body: TextAnnotationIn,
): Promise<TextAnnotation> {
  return apiFetch(`/language/texts/${textId}/annotations`, textAnnotationSchema, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updateTextAnnotation(
  id: number,
  body: Partial<TextAnnotationIn>,
): Promise<TextAnnotation> {
  return apiFetch(`/language/annotations/${id}`, textAnnotationSchema, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export async function deleteTextAnnotation(id: number): Promise<void> {
  const res = await fetch(`${API_URL}/language/annotations/${id}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new ApiError(res.status, body || res.statusText);
  }
}

export function lookupTextSelection(
  textId: number,
  selectedText: string,
): Promise<TextLookup> {
  return apiFetch(`/language/texts/${textId}/lookup`, textLookupSchema, {
    method: "POST",
    body: JSON.stringify({ selected_text: selectedText }),
  });
}

export function createCardFromAnnotation(
  annotationId: number,
  body: {
    deck_id: number;
    front?: string;
    back?: string;
    card_type?: CardType;
    direction?: CardDirection;
    tags?: string[];
    enrich?: boolean;
  },
): Promise<Card> {
  return apiFetch(`/language/annotations/${annotationId}/card`, cardSchema, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

// ---------------------------------------------------------------------------
// speak - generic tap-to-hear
// ---------------------------------------------------------------------------

/** Cached TTS URL for arbitrary text; "" when the server can't synthesize. */
export function fetchSpeechUrl(
  language: LanguageCode,
  text: string,
): Promise<string> {
  return apiFetch("/language/speak", z.object({ audio_url: z.string() }), {
    method: "POST",
    body: JSON.stringify({ language, text }),
  }).then((res) => res.audio_url);
}

// ---------------------------------------------------------------------------
// wiki
// ---------------------------------------------------------------------------

export const wikiSenseSchema = z.object({
  definition: z.string(),
  examples: z.array(z.string()),
  tags: z.array(z.string()),
  translation: z.string(),
});

export const wikiDictEntrySchema = z.object({
  part_of_speech: z.string(),
  ipa: z.string(),
  gender: z.string(),
  senses: z.array(wikiSenseSchema),
  synonyms: z.array(z.string()),
});

export const wikiConjugationSchema = z.object({
  mood: z.string(),
  tense: z.string(),
  person: z.string(),
  form: z.string(),
});
export type WikiConjugation = z.infer<typeof wikiConjugationSchema>;

export const wikiOutSchema = z.object({
  word: z.string(),
  language: languageCodeSchema,
  found: z.boolean(),
  source: z.string(),
  translated_from: z.string(),
  entries: z.array(wikiDictEntrySchema),
  gender: z.string(),
  ipa: z.string(),
  frequency: z.number().nullable(),
  lemma: z.string(),
  headword: z.string(),
  is_inflected: z.boolean(),
  form_note: z.string(),
  lexeme_id: z.number().nullable(),
  existing_cards: z.array(cardLocationSchema),
  cognate_note: z.string(),
  is_false_friend: z.boolean(),
  is_verb: z.boolean(),
  verb_id: z.number().nullable(),
  conjugations: z.array(wikiConjugationSchema),
  can_conjugate: z.boolean(),
  predicted: z.boolean(),
  equivalent: z
    .object({
      language: languageCodeSchema,
      word: z.string(),
      verb_id: z.number().nullable(),
    })
    .nullable(),
});
export type WikiResult = z.infer<typeof wikiOutSchema>;

/** Language the definitions are shown in - English or machine-translated. */
export type WikiDefsLanguage = "en" | LanguageCode;

export function wikiLookup(
  language: LanguageCode,
  word: string,
  defs: WikiDefsLanguage = "en",
): Promise<WikiResult> {
  return apiFetch(
    `/language/wiki/${language}/${encodeURIComponent(word)}?defs=${defs}`,
    wikiOutSchema,
  );
}

// ---------------------------------------------------------------------------
// conjugation center
// ---------------------------------------------------------------------------

export const verbSchema = z.object({
  id: z.number(),
  language: languageCodeSchema,
  lexeme_id: z.number().nullable(),
  infinitive: z.string(),
  group: z.string(),
  is_irregular: z.boolean(),
  translation: z.string(),
  frequency_rank: z.number(),
  equivalent_verb_id: z.number().nullable(),
  equivalent_language: z.string(),
  equivalent_infinitive: z.string(),
  set_ids: z.array(z.number()),
});
export type Verb = z.infer<typeof verbSchema>;

export const conjugationSchema = z.object({
  id: z.number(),
  mood: z.string(),
  tense: z.string(),
  person: z.string(),
  form: z.string(),
  slot_key: z.string(),
  equivalent_form: z.string(),
  audio_url: z.string(),
});
export type Conjugation = z.infer<typeof conjugationSchema>;

export const verbDetailSchema = verbSchema.extend({
  conjugations: z.array(conjugationSchema),
});
export type VerbDetail = z.infer<typeof verbDetailSchema>;

export function listVerbs(language: LanguageCode = "fr"): Promise<Verb[]> {
  return apiFetch(`/language/verbs?language=${language}`, z.array(verbSchema));
}

/** Conjugate + save a wiki verb into the conjugation center (verbecc). */
export function saveVerb(
  language: LanguageCode,
  infinitive: string,
  setIds: number[] = [],
): Promise<VerbDetail> {
  return apiFetch("/language/verbs", verbDetailSchema, {
    method: "POST",
    body: JSON.stringify({ language, infinitive, set_ids: setIds }),
  });
}

export function getVerb(id: number): Promise<VerbDetail> {
  return apiFetch(`/language/verbs/${id}`, verbDetailSchema);
}

export function getConjugationAudio(
  id: number,
): Promise<{ id: number; audio_url: string }> {
  return apiFetch(
    `/language/conjugations/${id}/audio`,
    z.object({ id: z.number(), audio_url: z.string() }),
    { method: "POST" },
  );
}

export type DrillIn = {
  mood: string;
  tense: string;
  verb_ids?: number[];
  group?: string;
  irregular_only?: boolean;
  audio_first?: boolean;
  language?: LanguageCode;
  set_id?: number;
};

export function createDrills(body: DrillIn): Promise<Deck> {
  return apiFetch("/language/drills", deckSchema, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export const verbSetSchema = z.object({
  id: z.number(),
  language: languageCodeSchema,
  name: z.string(),
  description: z.string(),
  verb_count: z.number(),
  created_at: z.string(),
});
export type VerbSet = z.infer<typeof verbSetSchema>;

export function listVerbSets(language?: LanguageCode): Promise<VerbSet[]> {
  const suffix = language ? `?language=${language}` : "";
  return apiFetch(`/language/verb-sets${suffix}`, z.array(verbSetSchema));
}

export function createVerbSet(body: {
  language: LanguageCode;
  name: string;
  description?: string;
}): Promise<VerbSet> {
  return apiFetch("/language/verb-sets", verbSetSchema, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function addVerbToSet(setId: number, verbId: number): Promise<VerbSet> {
  return apiFetch(`/language/verb-sets/${setId}/members`, verbSetSchema, {
    method: "POST",
    body: JSON.stringify({ verb_id: verbId }),
  });
}

export async function removeVerbFromSet(
  setId: number,
  verbId: number,
): Promise<void> {
  const res = await fetch(
    `${API_URL}/language/verb-sets/${setId}/members/${verbId}`,
    { method: "DELETE", credentials: "include" },
  );
  if (!res.ok) throw new ApiError(res.status, await res.text());
}

// ---------------------------------------------------------------------------
// bulk import - paste a word list, or upload a Kobo highlight database
// ---------------------------------------------------------------------------

export const importItemSchema = z.object({
  selected_text: z.string(),
  book: z.string(),
  front: z.string(),
  back: z.string(),
  ipa: z.string(),
  gender: z.string(),
  part_of_speech: z.string(),
  cognate_note: z.string(),
  is_false_friend: z.boolean(),
  is_inflected: z.boolean(),
  form_note: z.string(),
  lexeme_id: z.number().nullable(),
  existing_decks: z.array(z.string()),
});
export type ImportItem = z.infer<typeof importItemSchema>;

export const importPreviewSchema = z.object({
  language: languageCodeSchema,
  total_highlights: z.number(),
  items: z.array(importItemSchema),
});
export type ImportPreview = z.infer<typeof importPreviewSchema>;

export type ImportSource = "paste" | "kobo";

/** Resolve a pasted newline/comma-separated word list into card drafts. */
export function pasteImportPreview(
  language: LanguageCode,
  text: string,
): Promise<ImportPreview> {
  return apiFetch("/language/import/paste", importPreviewSchema, {
    method: "POST",
    body: JSON.stringify({ language, text }),
  });
}

/** Upload a KoboReader.sqlite; get back resolved, deduped card drafts. */
export async function koboImportPreview(
  language: LanguageCode,
  file: File,
): Promise<ImportPreview> {
  const form = new FormData();
  form.append("language", language);
  form.append("file", file);
  const res = await fetch(`${API_URL}/language/import/kobo`, {
    method: "POST",
    credentials: "include",
    body: form,
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new ApiError(res.status, body || res.statusText);
  }
  return importPreviewSchema.parse(await res.json());
}

export type ImportCommitCard = {
  front: string;
  back?: string;
  ipa?: string;
  gender?: string;
  part_of_speech?: string;
  cognate_note?: string;
  is_false_friend?: boolean;
  source_ref?: string;
};

export const importCommitSchema = z.object({
  created: z.number(),
  skipped: z.number(),
});
export type ImportCommitResult = z.infer<typeof importCommitSchema>;

export function importCommit(
  deckId: number,
  source: ImportSource,
  cards: ImportCommitCard[],
): Promise<ImportCommitResult> {
  return apiFetch("/language/import/commit", importCommitSchema, {
    method: "POST",
    body: JSON.stringify({ deck_id: deckId, source, cards }),
  });
}

File: backend/app/schemas/language.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models import CardDirection, CardSource, CardType, Language, ReviewStateName

# ---------------------------------------------------------------------------
# decks
# ---------------------------------------------------------------------------


class DeckOut(BaseModel):
    id: int
    name: str
    language: Language
    description: str
    tags: list[str]
    is_system: bool
    created_at: datetime
    card_count: int
    due_count: int
    new_count: int


class DeckCreate(BaseModel):
    name: str
    language: Language
    description: str = ""
    tags: list[str] = []

    @field_validator("name")
    @classmethod
    def name_not_blank(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("deck name is required")
        return value


class DeckUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    tags: list[str] | None = None


# ---------------------------------------------------------------------------
# cards
# ---------------------------------------------------------------------------


class CardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    deck_id: int
    card_type: CardType
    direction: CardDirection
    front: str
    back: str
    ipa: str
    gender: str
    part_of_speech: str
    audio_url: str
    example: str
    example_translation: str
    cognate_note: str
    is_false_friend: bool
    source: CardSource
    source_ref: str
    tags: list[str]
    created_at: datetime
    # flattened review state
    state: ReviewStateName
    due: datetime
    reps: int
    lapses: int
    stability: float | None


class CardCreate(BaseModel):
    deck_id: int
    front: str
    back: str = ""
    card_type: CardType = CardType.basic
    direction: CardDirection = CardDirection.recognition
    ipa: str = ""
    gender: str = ""
    part_of_speech: str = ""
    example: str = ""
    example_translation: str = ""
    cognate_note: str = ""
    tags: list[str] = []
    enrich: bool = True

    @field_validator("front")
    @classmethod
    def front_not_blank(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("card front is required")
        return value


class CardUpdate(BaseModel):
    front: str | None = None
    back: str | None = None
    card_type: CardType | None = None
    direction: CardDirection | None = None
    ipa: str | None = None
    gender: str | None = None
    part_of_speech: str | None = None
    example: str | None = None
    example_translation: str | None = None
    cognate_note: str | None = None
    is_false_friend: bool | None = None
    tags: list[str] | None = None
    deck_id: int | None = None


# ---------------------------------------------------------------------------
# study
# ---------------------------------------------------------------------------


class StudyQueue(BaseModel):
    cards: list[CardOut]
    due_count: int
    new_count: int


class ReviewIn(BaseModel):
    card_id: int
    rating: int = Field(ge=1, le=4)


class ReviewOut(BaseModel):
    card_id: int
    state: ReviewStateName
    due: datetime
    # True when the card comes back later in this same session (learning steps)
    again_soon: bool


# ---------------------------------------------------------------------------
# dashboard
# ---------------------------------------------------------------------------


class LanguageSplit(BaseModel):
    fr: int
    es: int


class RetentionOut(BaseModel):
    fr: float | None
    es: float | None


class GuardrailOut(BaseModel):
    es_reviews_this_week: int
    floor: int


class HeatmapDay(BaseModel):
    day: date
    count: int


class TaskOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    text: str
    done: bool
    task_date: date | None
    recurrence: str
    template_id: int | None
    action_ref: str


class TaskCreate(BaseModel):
    text: str
    task_date: date | None = None
    recurrence: str = ""  # "" | "daily"
    action_ref: str = ""

    @field_validator("text")
    @classmethod
    def text_not_blank(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("task text is required")
        return value


class TaskUpdate(BaseModel):
    text: str | None = None
    done: bool | None = None
    action_ref: str | None = None


class TargetOut(BaseModel):
    id: int
    metric: str
    label: str
    target: int
    auto: bool
    current: int


class TargetCreate(BaseModel):
    metric: str
    label: str
    target: int = Field(gt=0)
    auto: bool = True


class TargetUpdate(BaseModel):
    label: str | None = None
    target: int | None = Field(default=None, gt=0)


class LanguageDashboard(BaseModel):
    streak_days: int
    reviews_today: int
    due_today: LanguageSplit
    new_cards: LanguageSplit
    total_cards: LanguageSplit
    mature_cards: LanguageSplit
    retention_30d: RetentionOut
    guardrail: GuardrailOut
    tasks: list[TaskOut]
    targets: list[TargetOut]
    heatmap: list[HeatmapDay]
    decks: list[DeckOut]


# ---------------------------------------------------------------------------
# text library + annotations
# ---------------------------------------------------------------------------


class TextAnnotationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    text_id: int
    start_offset: int
    end_offset: int
    selected_text: str
    kind: str
    color: str
    note: str
    translation: str
    ipa: str
    part_of_speech: str
    cognate_note: str
    is_false_friend: bool
    deck_id: int | None
    flashcard_id: int | None
    created_at: datetime
    updated_at: datetime


class LanguageTextOut(BaseModel):
    id: int
    title: str
    language: Language
    source_type: str
    source_ref: str
    tags: list[str]
    created_at: datetime
    updated_at: datetime
    annotation_count: int


class LanguageTextDetail(LanguageTextOut):
    content: str
    annotations: list[TextAnnotationOut]


class LanguageTextCreate(BaseModel):
    title: str
    language: Language
    content: str
    source_type: str = "other"
    source_ref: str = ""
    tags: list[str] = Field(default_factory=list)

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("title is required")
        return value

    @field_validator("content")
    @classmethod
    def content_not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("text content is required")
        return value


class LanguageTextUpdate(BaseModel):
    title: str | None = None
    language: Language | None = None
    content: str | None = None
    source_type: str | None = None
    source_ref: str | None = None
    tags: list[str] | None = None


class TextAnnotationCreate(BaseModel):
    start_offset: int = Field(ge=0)
    end_offset: int = Field(gt=0)
    selected_text: str = ""
    kind: str = "highlight"
    color: str = "brand"
    note: str = ""
    translation: str = ""
    ipa: str = ""
    part_of_speech: str = ""
    cognate_note: str = ""
    is_false_friend: bool = False


class TextAnnotationUpdate(BaseModel):
    start_offset: int | None = Field(default=None, ge=0)
    end_offset: int | None = Field(default=None, gt=0)
    selected_text: str | None = None
    kind: str | None = None
    color: str | None = None
    note: str | None = None
    translation: str | None = None
    ipa: str | None = None
    part_of_speech: str | None = None
    cognate_note: str | None = None
    is_false_friend: bool | None = None


class TextLookupIn(BaseModel):
    selected_text: str

    @field_validator("selected_text")
    @classmethod
    def selected_text_not_blank(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("selected text is required")
        return value


class TextLookupOut(BaseModel):
    selected_text: str
    translation: str
    ipa: str
    part_of_speech: str
    cognate_note: str
    is_false_friend: bool
    provider: str


class AnnotationCardCreate(BaseModel):
    deck_id: int
    front: str = ""
    back: str = ""
    card_type: CardType = CardType.basic
    direction: CardDirection = CardDirection.recognition
    tags: list[str] = Field(default_factory=list)
    enrich: bool = True


# ---------------------------------------------------------------------------
# conjugation center
# ---------------------------------------------------------------------------


class VerbOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    infinitive: str
    group: str
    is_irregular: bool
    translation: str
    es_equivalent: str
    frequency_rank: int


class ConjugationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    mood: str
    tense: str
    person: str
    form: str
    es_form: str
    audio_url: str


class VerbDetail(VerbOut):
    conjugations: list[ConjugationOut]


class ConjugationAudioOut(BaseModel):
    id: int
    audio_url: str


class DrillCreate(BaseModel):
    """Generate cell-level cloze cards into a per-tense system deck."""

    mood: str
    tense: str
    verb_ids: list[int] = []  # empty = use group/irregular filters
    group: str = ""  # "-er" | "-ir" | "-re" | "irregular" | ""
    irregular_only: bool = False
    audio_first: bool = False
